# MovieLens 32M - Exploratory Data Analysis

## AuraCinema - Movie Discovery Engine

This notebook is a data analysis project which explores the MovieLens 32M dataset as the data foundation that will be used further in a Machine Learning project: "AuraCinema Movie Recommendation System", an AI-powered application that produces personalised movie recommendations based in user's preferences.


## 1. Introduction

The objective of this exploratory data analysis (EDA) is to understand the structure, quality and characteristics of the MovieLens 32M dataset before developing the movie recommendation system.

The analysis will investigate the available data, user and movie activity, rating behaviour, movie genres and user–movie interactions.

The findings from this analysis will be used to guide subsequent data preprocessing and the design of the recommendation approach.

## 2. Dataset Overview

MovieLens 32M is a stable benchmark dataset provided by GroupLens. It contains movie ratings, movie metadata, user-generated tags and links to external movie databases.

The dataset will be used as the primary data source for developing and evaluating the recommendation system.

### 2.1 Dataset Components

The MovieLens 32M dataset is composed of four main data files:

| File | Description |
|---|---|
| `ratings.csv` | User ratings for movies |
| `movies.csv` | Movie titles and genre information |
| `tags.csv` | User-generated tags associated with movies |
| `links.csv` | Links between MovieLens movie IDs and external movie databases |

The `ratings.csv` file represents the main interaction data for the recommendation problem, while the other files provide additional information that may support data exploration and future recommendation features.

In [5]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

In [6]:
data_path = "../data/ml-32m/"

In [7]:
movies = pd.read_csv(data_path + "movies.csv")

In [9]:
movies.head()

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


In [17]:
print('How many movies in the catalog?\n')

print(f'The catalog has {movies.shape[0]} movies, and {movies.shape[1]} main attributes, which are: \n\n{movies.columns}.')

How many movies in the catalog?

The catalog has 87585 movies, and 3 main attributes, which are: 

Index(['movieId', 'title', 'genres'], dtype='str').


In [18]:
print('Getting information about the dataset:\n')

movies.info()

Getting information about the dataset:

<class 'pandas.DataFrame'>
RangeIndex: 87585 entries, 0 to 87584
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   movieId  87585 non-null  int64
 1   title    87585 non-null  str  
 2   genres   87585 non-null  str  
dtypes: int64(1), str(2)
memory usage: 2.0 MB


### Initial observations

The `movies.csv` file contains 87,585 movie records and three columns: `movieId`, `title` and `genres`.

The `movieId` column provides the identifier used to connect movies with the other MovieLens files. The `title` column contains the movie title, including the release year when available. The `genres` column contains one or more genre labels for each movie.

### Notes on the ratings dataset:
 
The `ratings.csv` file contains over 32 million records. To avoid unnecessarily loading the complete dataset during the initial inspection, a sample of 10,000 records is used at this stage.

The full ratings dataset will be loaded or processed in chunks when required for subsequent analysis.

In [24]:
ratings_sample = pd.read_csv(data_path + 'ratings.csv', nrows=10000)

ratings_sample.head()

,userId,movieId,rating,timestamp
0,1,17,4.0,944249077
1,1,25,1.0,944250228
2,1,29,2.0,943230976
3,1,30,5.0,944249077
4,1,32,5.0,943228858


In [21]:
ratings_sample.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   userId     10000 non-null  int64  
 1   movieId    10000 non-null  int64  
 2   rating     10000 non-null  float64
 3   timestamp  10000 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 312.6 KB


In [23]:
tags = pd.read_csv(data_path + 'tags.csv')

tags.head()

,userId,movieId,tag,timestamp
0,22,26479,Kevin Kline,1583038886
1,22,79592,misogyny,1581476297
2,22,247150,acrophobia,1622483469
3,34,2174,music,1249808064
4,34,2174,weird,1249808102


In [25]:
links = pd.read_csv(data_path + 'links.csv')

links.head()

,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


### Brief dataset overview:

In [26]:
dataset_overview = pd.DataFrame({
    "Dataset": ["Movies", "Ratings sample", "Tags", "Links"],
    "Rows": [
        len(movies),
        len(ratings_sample),
        len(tags),
        len(links)
    ],
    "Columns": [
        movies.shape[1],
        ratings_sample.shape[1],
        tags.shape[1],
        links.shape[1]
    ]
})

dataset_overview

,Dataset,Rows,Columns
0,Movies,87585,3
1,Ratings sample,10000,4
2,Tags,2000072,4
3,Links,87585,3
